src/resolve/resolved-manager.c

```diff
@@ -196,6 +197,40 @@ static int manager_process_address(sd_netlink *rtnl, sd_netlink_message *mm, voi
         return 0;
 }
 
+static int manager_process_route(sd_netlink *rtnl, sd_netlink_message *mm, void *userdata) {
+        Manager *m = ASSERT_PTR(userdata);
+        Link *l;
+        uint16_t type;
+        int r;
+
+        assert(rtnl);
+        assert(mm);
+
+        r = sd_netlink_message_get_type(mm, &type);
+        if (r < 0) {
+                log_warning_errno(r, "Failed not get message type, ignoring: %m");
+                return 0;
+        }
+
+        if (!IN_SET(type, RTM_NEWROUTE, RTM_DELROUTE)) {
+                log_warning("Unexpected message type %u when processing route, ignoring.", type);
+                return 0;
+        }
+
+        /* Routing table changes may affect the accessibility of DNS servers. */
+        HASHMAP_FOREACH(l, m->links) {
+                LIST_FOREACH(servers, s, l->dns_servers)
+                        dns_server_reset_accessible(s);
+        }
+
+        LIST_FOREACH(servers, s, m->dns_servers)
+                dns_server_reset_accessible(s);
+
+        (void) manager_send_dns_configuration_changed(m);
+
```

Comment:
so i am a bit concerned that we'll run through this for a myriad of routes (because plenty of systems have a *lot* of routes, and expect resolved to deal with this). if the system also has a lot of interfaces this is a lot of work, and that over and over again. and this might send out IPC notifications, which is even worse

one idea: move this into a defer job. i.e. and sd_event_add_defer() job that runs at a low priority and is enabled whenever we see RTM_NEWROUTE/RTM_DELROUTE. that defer job would then do the actual work, when the event loop is idle again

Reply:
also, maybe extract the ifindex from the route msg (if there is one), and only invalidate the accessible flag of the relevant interfaces (and of the global servers).

just to reduce impact

Reply:
another idea: maybe only enable the RTM_NEWROUTE/RTM_DELROUTE matches while at least one client cares? and disable them when the last one is gone? 

that would substantially reduce the chance that people start complaining about this

Reply:
I added logic to only enable this match when there is at least one client connected. I also added logic to only emit the notification when the resulting JSON has changed, in an attempt to reduce IPC notifications related to this.